File: jun1/src/active_combinator.py
from typing import Optional

from .combinator import Combinator
from .parse_iteration_result import FrameStack, ParserIterationResult


class ActiveCombinator:
    def __init__(self, combinator: Combinator, names: Optional[list[str]] = None) -> None:
        self.combinator = combinator
        self.signal_id = 0
        if names is not None:
            self.frame_stack = FrameStack()
            for name in names:
                self.frame_stack.push_name(name.encode())
            self.state = combinator.initial_state(self.signal_id, self.frame_stack)
        else:
            self.state = combinator.initial_state(self.signal_id, FrameStack())

    def send(self, c: Optional[str]) -> ParserIterationResult:
        self.state, self.signal_id, result = self.combinator.next_state(self.state, c, self.signal_id)
        return result

File: jun1/src/active_combinator.rs
use std::rc::Rc;
use crate::combinator::Combinator;
use crate::parse_iteration_result::{FrameStack, ParserIterationResult};
use crate::state::CombinatorState;

pub struct ActiveCombinator<C, State> {
    combinator: C,
    state: State,
    signal_id: usize,
}

impl<C, State> ActiveCombinator<C, State>
where
    C: Combinator<State = State>,
{
    pub fn new(combinator: C) -> Self {
        let mut signal_id = 0;
        let state = combinator.initial_state(&mut signal_id, FrameStack::default());
        Self {
            combinator,
            state,
            signal_id,
        }
    }

    pub fn new_with_names(combinator: C, names: Vec<String>) -> Self {
        let mut signal_id = 0;
        let mut frame_stack = FrameStack::default();
        for name in names {
            frame_stack.push_name(name.as_bytes());
        }
        let state = combinator.initial_state(&mut signal_id, frame_stack);
        Self {
            combinator,
            state,
            signal_id,
        }
    }

    pub fn send(&mut self, c: Option<char>) -> ParserIterationResult {
        self.combinator
            .next_state(&mut self.state, c, &mut self.signal_id)
    }
}
